File: Prosary/Models/RosaryOptions.swift

//
//  RosaryOptions.swift
//  Prosary
//
//  Configuration options specific to the Rosary. Lives inside a Prayer when kind == .rosary.
//

import Foundation

struct RosaryOptions: Hashable, Codable {
  var mysterySelectionMode: MysterySelectionMode = .todaysMysteries

  /// Used only when `mysterySelectionMode` is `.specific`.
  var specificMysteryGroup: MysteryGroup = .joyful

  var includeApostlesCreed: Bool = true

  /// The opening Our Father + 3 Hail Marys (for faith, hope, and charity) + Glory Be.
  var includeOpeningPrayers: Bool = true

  /// The Fatima Prayer ("O my Jesus...") recited after the Glory Be of each decade.
  var includeFatimaPrayer: Bool = true

  var eternalRestForDeceased: EternalRestPlacement = .none

  var marianAntiphon: MarianAntiphonOption = .seasonal

  var includeStMichaelPrayer: Bool = false

  var includeFinalSignOfCross: Bool = true

  var mysterySelectionSummary: String {
    switch mysterySelectionMode {
    case .specific:
      return String(localized: "rosaryOptions.summary.always", defaultValue: "Always \(specificMysteryGroup.displayName)")
    case .fifteenMystery:
      return String(localized: "rosaryOptions.summary.fifteenMystery", defaultValue: "The 15 Mysteries")
    case .twentyMystery:
      return String(localized: "rosaryOptions.summary.twentyMystery", defaultValue: "The 20 Mysteries")
    case .todaysMysteries:
      return String(localized: "mysterySelectionMode.todaysMysteries", defaultValue: "Today's Mysteries")
    }
  }
}

File: Prosary/Models/RosaryOptions.cs
namespace Prosary.Models;

/// <summary>Configuration options specific to the Rosary. Lives inside a <see cref="Prayer"/>
/// when <c>Kind == PrayerKind.Rosary</c>.</summary>
public sealed record RosaryOptions
{
    public MysterySelectionMode MysterySelectionMode { get; init; } = MysterySelectionMode.TodaysMysteries;

    /// <summary>Used only when <see cref="MysterySelectionMode"/> is <see cref="Models.MysterySelectionMode.Specific"/>.</summary>
    public MysteryGroup SpecificMysteryGroup { get; init; } = MysteryGroup.Joyful;

    public bool IncludeApostlesCreed { get; init; } = true;

    /// <summary>The opening Our Father + 3 Hail Marys (for faith, hope, and charity) + Glory Be.</summary>
    public bool IncludeOpeningPrayers { get; init; } = true;

    /// <summary>The Fatima Prayer ("O my Jesus...") recited after the Glory Be of each decade.</summary>
    public bool IncludeFatimaPrayer { get; init; } = true;

    public EternalRestPlacement EternalRestForDeceased { get; init; } = EternalRestPlacement.None;

    public MarianAntiphonOption MarianAntiphon { get; init; } = MarianAntiphonOption.Seasonal;

    public bool IncludeStMichaelPrayer { get; init; } = false;

    public bool IncludeFinalSignOfCross { get; init; } = true;

    public string MysterySelectionSummary => MysterySelectionMode switch
    {
        MysterySelectionMode.Specific => $"Always {SpecificMysteryGroup}",
        MysterySelectionMode.FifteenMystery => "The 15 Mysteries",
        MysterySelectionMode.TwentyMystery => "The 20 Mysteries",
        MysterySelectionMode.TodaysMysteries => "Today's Mysteries",
        _ => throw new ArgumentOutOfRangeException(nameof(MysterySelectionMode))
    };
}
